nova/compute/api.py

```diff
@@ -1773,10 +1774,15 @@ def soft_delete(self, context, instance):
         """Terminate an instance."""
         LOG.debug('Going to try to soft delete instance',
                   instance=instance)
-
-        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
-                     task_state=task_states.SOFT_DELETING,
-                     deleted_at=timeutils.utcnow())
+        if instance['vm_state'] == vm_states.SOFT_DELETED:
+            LOG.warn("Instance %s is soft delete state" % (instance.uuid))
+            reason = ("The instance is %s state" % (instance.vm_state))
+            raise exception.InstanceTerminationFailure(reason=reason)
+        else:
+            self._delete(context, instance, 'soft_delete',
+                         self._do_soft_delete,
+                         task_state=task_states.SOFT_DELETING,
+                         deleted_at=timeutils.utcnow())
 
```

Comment:
这里不只是配额的问题，还涉及到测试时，如果第一次删除没有提示信息，你是否认为已经成功了？
基于uuid执行第二次删除依旧没有提示信息，你是否认为还是成功了？ 那你这样一直执行下去，你是否认为一直都成功了？你认为这是否是一个BUG ?
这个地方之所以不用上游的删除后，在进行配额回滚，是从根本上就不允许，已经处于软删除状态的云主机再次执行软删除操作．

2016-07-08 11:51 GMT+08:00 daemon <notifications@github.com>:

> In nova/compute/api.py
> <https://github.com/eayunstack/nova/pull/37#discussion_r70022488>:
>
> > @@ -1773,10 +1774,15 @@ def soft_delete(self, context, instance):
> >          """Terminate an instance."""
> >          LOG.debug('Going to try to soft delete instance',
> >                    instance=instance)
> > -
> > -        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
> > -                     task_state=task_states.SOFT_DELETING,
> > -                     deleted_at=timeutils.utcnow())
> > +        if instance['vm_state'] == vm_states.SOFT_DELETED:
> > +            LOG.warn("Instance %s is soft delete state" % (instance.uuid))
> > +            reason = ("The instance is %s state" % (instance.vm_state))
> > +            raise exception.InstanceTerminationFailure(reason=reason)
> > +        else:
> > +            self._delete(context, instance, 'soft_delete',
> > +                         self._do_soft_delete,
> > +                         task_state=task_states.SOFT_DELETING,
> > +                         deleted_at=timeutils.utcnow())
> >
>
> 我觉得既然是关于配额处理的问题，最好放到_delete里面，与其他关于配额处理的代码放一块。可以在
> https://git.openstack.org/cgit/openstack/nova/commit/?id=4546ce0012b526cd370533ba5a3f2f4107d249a2的
> 基础上加以修改
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eayunstack/nova/pull/37/files/b9f59ddec6dfb4785ef13559f31c37d1239b21e7#r70022488>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKMiZudlsCaMfcjfycbWnoKed1zd1nipks5qTclJgaJpZM4I0Frh>
> .
>